### PR TITLE
go.*,pkg,vendor: Bump controller-runtime to v0.10.1

### LINKED ISSUE #2353: Remove controller-runtime replace pin
## Feature Request

**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Example: "I have an issue when (...)"

Remove the [replace pin](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/go.mod#L66) for controller runtime once the race condition issue around metadata-only watchers has been resolved.

**Describe the solution you'd like**
A clear and concise description of what you want to happen. Add any considered drawbacks.

Bump the controller runtime dependency and remove the replace pin once https://github.com/kubernetes-sigs/controller-runtime/pull/1650 gets tagged in a release.

### PR DESCRIPTION
**Pull request by timflannagan**
**Description of the change:**
Bumps the controller runtime dependency to v0.10.1 which contains the race condition fix around metadata-only watchers.

**Motivation for the change:**
Removes the replace pin in the root go.mod.

**Reviewer Checklist**
- [x] Implementation matches the proposed design, or proposal is updated to match implementation
- [x] Sufficient unit test coverage
- [x] Sufficient end-to-end test coverage
- [x] Docs updated or added to `/doc`
- [x] Commit messages sensible and descriptive

Fixes #2353

### CODE REVIEW

File: test/e2e/deprecated_e2e_test.go [hunk 1]
```diff
@@ -30,15 +30,15 @@ var _ = Describe("Not found APIs", func() {
 		// each entry is an installplan with a deprecated resource
 		type payload struct {
 			name       string
-			ip         *operatorsv1alpha1.InstallPlan
+			IP         *operatorsv1alpha1.InstallPlan
```
Comment (awgreene) [inline test/e2e/deprecated_e2e_test.go:33]: Non-blocking: Why was this exported?

**Comment by awgreene** [comment]
```
[BeforeEach] Operator API
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:41
[It] should surface components in its status
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:73
STEP: eventually having a status that contains its component label selector
STEP: eventually listing a single component reference
first patch error: Operation cannot be fulfilled on serviceaccounts "sa-a-gcg2h": the object has been modified; please apply your changes to the latest version and try again
STEP: eventually listing multiple component references
```
Is there potentially a race condition [here](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/test/e2e/operator_test.go#L184)?

File: pkg/controller/operators/suite_test.go [hunk 2]
```diff
@@ -88,14 +88,14 @@ var _ = BeforeSuite(func() {
...
 		UseExistingCluster: &useExisting,
-		CRDs: []client.Object{
-			crds.CatalogSource(),
-			crds.ClusterServiceVersion(),
-			crds.InstallPlan(),
-			crds.Subscription(),
-			crds.OperatorGroup(),
-			crds.Operator(),
-			crds.OperatorCondition(),
+		CRDs: []apiextensionsv1.CustomResourceDefinition{
```
Comment (tylerslaton) [inline pkg/controller/operators/suite_test.go:91]: Is this something we should move to across the board when using controller-runtime moving forward?

**Comment by timflannagan** [comment]
Holding so the bot doesn't go crazy.

/hold

**Review by timflannagan — COMMENTED** [review]
It seems like we're continuously running into the following errors:

```
Summarizing 2 Failures:

[Fail] Operator API [It] should surface components in its status 
/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/util_test.go:838

[Fail] Operator API when a subscription to a package exists [It] should automatically adopt components 
/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:346
```

When poking around these errors locally, it seems like both of the test failures are around existing ServiceAccount's that aren't being propagated to the list of status.components.refs array. Maybe we're missing that cache event entirely in some cases?

**Comment by timflannagan** [comment]
Hmm weird it looks like we do get the mapping function event properly but we're somehow not able to update the Operator resource:

```
2021-12-08T19:40:37.265148583Z stderr F 2021-12-08T19:40:37.265Z	INFO	controllers.adoption	mapToClusterServiceVersions: requeuing a csv	{"gvk": "&TypeMeta{Kind:,APIVersion:,}", "name": "kiali-operator", "namespace": "ns-4f8q4"}
2021-12-08T19:40:37.265357386Z stderr F 2021-12-08T19:40:37.265Z	INFO	controllers.operator	mapComponentRequests: requeuing operator resource	{"gvk": "&TypeMeta{Kind:,APIVersion:,}", "name": "kiali-operator", "namespace": "ns-4f8q4"}
```

I'm going to try and update to also collect the state of that Operator resource.

**Comment by estroz** [comment]
@timflannagan hoping #2515 doesn't screw this PR up. @dinhxuanvu and I needed to get unblocked for constraint stuff added to api.

**Comment by timflannagan** [comment]
@estroz Let's definitely touch base tomorrow. We'll need to cut an o-f/api minor release in any case, but there's some complexity involved with certain controller runtime versions, and it's compatibility with OLM.

cc @awgreene @njhale

File: go.mod [hunk 3]
```diff
@@ -41,28 +41,29 @@ require (
...
 replace (
 	// controller runtime
-	github.com/openshift/api => github.com/openshift/api v0.0.0-20200331152225-585af27e34fd // release-4.5
+	github.com/openshift/api => github.com/openshift/api v0.0.0-20211014063134-be2a7fb8aa44
 	github.com/openshift/client-go => github.com/openshift/client-go v0.0.0-20200326155132-2a6cd50aedd0 // release-4.5
+
 	// Patch for a race condition involving metadata-only
 	// informers until it can be resolved upstream:
-	sigs.k8s.io/controller-runtime v0.9.2 => github.com/benluddy/controller-runtime v0.9.3-0.20210720171926-9bcb99bd9bd3
+	sigs.k8s.io/controller-runtime v0.10.0 => github.com/timflannagan/controller-runtime v0.10.1-0.20211210161403-6756a4203e70
```
Comment (timflannagan) [inline go.mod:68]: Tracking the removal of this pin in https://github.com/operator-framework/operator-lifecycle-manager/issues/2353#event-5749763881.

**Pull request merged**